internal/controller/nonadmindownloadrequest_controller.go

```diff
@@ -0,0 +1,292 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package controller contains all controllers of the project
+package controller
+
+import (
+	"context"
+	"time"
+
+	velerov1 "github.com/vmware-tanzu/velero/pkg/apis/velero/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/util/workqueue"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	ctrlpredicate "sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	nacv1alpha1 "github.com/migtools/oadp-non-admin/api/v1alpha1"
+	"github.com/migtools/oadp-non-admin/internal/common/constant"
+	"github.com/migtools/oadp-non-admin/internal/common/function"
+)
+
+// NonAdminDownloadRequestReconciler reconciles a NonAdminDownloadRequest object
+type NonAdminDownloadRequestReconciler struct {
+	client.Client
+	Scheme        *runtime.Scheme
+	OADPNamespace string
+}
+
+const statusPatchErr = "unable to patch status condition"
+
+// +kubebuilder:rbac:groups=oadp.openshift.io,resources=nonadmindownloadrequests,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=oadp.openshift.io,resources=nonadmindownloadrequests/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=oadp.openshift.io,resources=nonadmindownloadrequests/finalizers,verbs=update
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// TODO(user): Modify the Reconcile function to compare the state specified by
+// the NonAdminDownloadRequest object against the actual cluster state, and then
+// perform operations to make the cluster state reflect the state specified by
+// the user.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.20.2/pkg/reconcile
+//
+// This reconcile implements ObjectReconciler https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.20.2/pkg/reconcile#ObjectReconciler
+// Each reconciliation event gets the associated object from Kubernetes before passing it to Reconcile
+func (r *NonAdminDownloadRequestReconciler) Reconcile(ctx context.Context, req *nacv1alpha1.NonAdminDownloadRequest) (reconcile.Result, error) {
```

Comment:
I know it may be a lot to ask, but were you considering following other controllers logical functions instead of one big chunk of reconcile code? Such as:
  https://github.com/migtools/oadp-non-admin/blob/master/internal/controller/nonadminbackup_controller.go#L110-L164

Reply:
Let me know which part is hard to read in this code and I'll make it a function.

Reply:
there are many reconcile many conditional branches and returns from them within the block of the code, so for me it was hard to follow in which state happens what. 13 different return statements within reconcile make hard to understand what is the prior to return based on the all previous ones condition.

Reply:
I would logically do:
```go
func (r *NonAdminDownloadRequestReconciler) handleExpiredReques
	// Logic for handling expired requests
	
func (r *NonAdminDownloadRequestReconciler) updateVeleroDownloadRequestStatus
	// Logic for updating Velero download request status

func (r *NonAdminDownloadRequestReconciler) processTargetKind
	// Logic for processing different target kinds
```